File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/services/FLTStatisticsService.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimlib.sdk.NIMClient
import com.netease.nimlib.sdk.v2.statistics.V2NIMStatisticsService
import com.netease.yunxin.kit.alog.ALog
import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * flutter 统计信息组件
 *
 * @constructor
 *
 * @param applicationContext
 * @param nimCore
 */
class FLTStatisticsService(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {

    private val tag = "FLTStatisticsService"

    override val serviceName = "StatisticsService"

    init {
        nimCore.onInitialized {
            registerFlutterMethodCalls(
                "getDatabaseInfos" to ::getDatabaseInfos
            )
        }
    }

    private suspend fun getDatabaseInfos(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMStatisticsService::class.java).getDatabaseInfos(
                { databaseInfoList ->
                    val result = databaseInfoList?.map { info ->
                        mapOf(
                            "path" to info.path,
                            "name" to info.name,
                            "size" to info.size
                        )
                    } ?: emptyList<Map<String, Any?>>()
                    ALog.i(tag, "getDatabaseInfos success, count=${result.size}")
                    cont.resume(
                        NimResult(
                            0,
                            data = mapOf("databaseInfoList" to result)
                        )
                    )
                },
                { error ->
                    ALog.e(tag, "getDatabaseInfos error: code=${error.code}, desc=${error.desc}")
                    cont.resume(NimResult(error.code, errorDetails = error.desc))
                }
            )
        }
    }
}

File: nim_core_v2/ios/Classes/services/FLTStatisticsService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import NIMSDK

enum StatisticsServiceMethodType: String {
  case getDatabaseInfos
}

class FLTStatisticsService: FLTBaseService, FLTService {
  func serviceName() -> String {
    ServiceType.StatisticsService.rawValue
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any],
                      _ resultCallback: ResultCallback) {
    switch method {
    case StatisticsServiceMethodType.getDatabaseInfos.rawValue:
      getDatabaseInfos(arguments, resultCallback)
    default:
      resultCallback.notImplemented()
    }
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  // MARK: - SDK API

  func getDatabaseInfos(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    weak var weakSelf = self
    NIMSDK.shared().v2StatisticsService.getDatabaseInfos { databaseInfoList in
      var result = [[String: Any]]()
      databaseInfoList.forEach { info in
        var item = [String: Any]()
        item["path"] = info.path
        item["name"] = info.name
        item["size"] = info.size
        result.append(item)
      }
      weakSelf?.successCallBack(resultCallback, ["databaseInfoList": result])
    } failure: { error in
      weakSelf?.errorCallBack(resultCallback, error.nserror.localizedDescription, Int(error.code))
    }
  }
}
